# Используем кластер postgres 15 из ДЗ №3
# Кластер развёрнут на ubuntu 20.04 в VirtualBox 7<br/>

## Подключимся клиентом:
## psql -U postgres -h localhost
### (запрашивается пароль, вводим его)
## Создадим базу для тестов:
## CREATE DATABASE test;
## Подключимся к ней:
## \c test
## Установим создание контрольной точки раз в 30 сек.:
## ALTER SYSTEM SET checkpoint_timeout = '30s';
## Перезагрузим конфиг:
## SELECT pg_reload_conf();
## Засечём позицию в WAL:
## SELECT pg_current_wal_lsn();
###  pg_current_wal_lsn
### ------------
###  1/76427DF0
### (1 row)
## Выйдем:
## \q
## Инициализируем pg_bench:
## pgbench -h 0.0.0.0 -i test
## Запустим 10-минутный тест:
## pgbench -h 0.0.0.0 -T 600 test
### tps = 463.745121
## Подключимся к базе:
## psql -U postgres -h localhost -d test
### (запрашивается пароль, вводим его)
## Посмотрим позицию в WAL:
## SELECT pg_current_wal_lsn();
###  pg_current_wal_lsn
### ------------
###  1/8E431F40
### (1 row)
## Вычислим объём сгенерированных логов:
## SELECT '1/8E431F40'::pg_lsn - '1/76427DF0'::pg_lsn;
###  ?column?
### -----------
###  402694480
### (1 row)
## Сгенерилось на 400 Мб. За 10 минут должно было пройти 20 контрольных точек - в среднем около 20 Мб на точку.
## Выйдем:
## \q
## Посмотрим логи:
## tail -n 100 /var/log/postgresql/postgresql-15-main.log
### Судя по логам, в моём случае все чекпоинты создавались строго по расписанию - в 11 и 41 секунду каждой минуты. Проверим это
## Подключимся:
## psql -U postgres -h localhost -d test
### (запрашивается пароль, вводим его)
## Запросим данные статистики:
## SELECT * FROM pg_stat_bgwriter \gx
### -[ RECORD 1 ]---------+------------------------------
### checkpoints_timed     | 49
### checkpoints_req       | 0
### checkpoint_write_time | 565504
### checkpoint_sync_time  | 324
### buffers_checkpoint    | 39054
### buffers_clean         | 0
### maxwritten_clean      | 0
### buffers_backend       | 1812
### buffers_backend_fsync | 0
### buffers_alloc         | 1805
### stats_reset           | 2025-02-06 23:29:01.072774+03
## Иногда при большой нагрузке БД не успевает и сама делает контрольные точки. Такие случаи регистрируются в поле checkpoints_req. Здесь, как мы видим, там 0
## Включим асинхронный коммит:
## ALTER SYSTEM SET synchronous_commit = off;
## SELECT pg_reload_conf();
## Запустим тест (минуты достаточно):
## pgbench -h 0.0.0.0 -T 60 test
### tps = 990.138639
## Tps больше, потому что сервер сообщает об успешном завершении сразу, как только транзакция будет завершена логически, прежде чем сгенерированные записи WAL фактически будут записаны на диск
## Выйдем:
## \q

## Созданим новый кластер с контрольными суммами:
## sudo pg_createcluster 15 second -- --data-checksums
## Запустим его:
## sudo pg_ctlcluster 15 second start
## Подключимся:
## psql -U postgres -h localhost -p 5433
### (запрашивается пароль, вводим его)
## Создадим базу для тестов:
## CREATE DATABASE test;
## Подключимся к ней:
## \c test
## Создадим таблицу:
## CREATE TABLE test_table(c1 text);
## Вставим записи:
## INSERT INTO test_table VALUES ('1'), ('2'), ('3');
## Определим путь, где лежит таблица:
## SELECT pg_relation_filepath('test_table');
### base/16388/16389
## Выйдем:
## \q
## Остановим кластер:
## sudo systemctl stop postgresql@15-second
## Отредактируем файл таблицы (заменим 3 символа в произвольном месте на 123):
## sudo nano /var/lib/postgresql/15/second/base/16388/16389
## Включим кластер:
## sudo systemctl start postgresql@15-second
## Подключимся к базе:
## psql -U postgres -h localhost -p 5433 -d test
### (запрашивается пароль, вводим его)
## Попытаемся выбрать данные из таблицы:
## SELECT * FROM test_table;
### WARNING:  page verification failed, calculated checksum 50206 but expected 57494
### ERROR:  invalid page in block 0 of relation base/16388/16389
## Таблицы существует, но она повреждена - контрольные суммы не бьются
## Включим параметр игнорирования контрольных сумм:
## ALTER SYSTEM SET ignore_checksum_failure = on;
## SELECT pg_reload_conf();
## Снова пробуем селектить:
## SELECT * FROM test_table;
### WARNING:  page verification failed, calculated checksum 50206 but expected 57494
###  c1 
### ----
###  1
###  2
###  3
### (3 rows)
## Предупреждение по-прежнему появляется, но таблица открывается - включенный параметр ignore_checksum_failure это позволяет
